### ISSUE TITLE
Eloquent create method casts timestamps to carbon when getdates returns empty array

### ISSUE THREAD

**Issue by mallinus** [laravel/framework#12148]
The Eloquent create method casts timestamps to Carbon instances even when overriding the getDates method to return an empty array.

**Comment by GrahamCampbell** [comment]
Set the timestamps property to false?

**Comment by mallinus** [comment]
Setting the timestamps to false will disable the automatic inserting and updating of the timestamps values. I do want to keep these timestamps updated. What I do not want is them to be casted to Carbon instances inserting models. I'm currently working on an API and these timestamps are casted to integers (in U format) via transformers so I do not need them to be casted. I've overriden the method getDates as suggested in the documentation and it seems to be working when fetching rows. However when inserting a model using the create method the timestamps are getting casted to Carbon instances.

**Comment by GrahamCampbell** [comment]
Are you sure you're overriding it correctly?

**Comment by mallinus** [comment]
`public function getDates()
    {
        return [];
    }`

As suggested in the documentation. I suppose this is the right way to do it since it's working fine for all cases except when creating?

**Comment by GrahamCampbell** [comment]
Ah, I see. Not a bug. It's because creating sets a date time instance.

**Comment by GrahamCampbell** [comment]
You'll need to look into mutators.

**Comment by mallinus** [comment]
I don't believe it would be good practice to cast types around when there's no need to in the first place, performance wise.

**Comment by GrahamCampbell** [comment]
The performance impact is minimal, compared to the actual work being done by the database to get your results out. Casting is `O(1)`.

**Comment by mallinus** [comment]
https://github.com/laravel/framework/blob/5.2/src/Illuminate/Database/Eloquent/Model.php#L1856

This freshTimestamp method is used when inserting to set updated_at and created_at attributes. It currently returns a Carbon instance. It should return a DateTime instance in case getDates() is an empty array.

**Comment by mallinus** [comment]
Upcasting and downcasting for no reason is still bad design even when the performance hit is minimal. I want my 40ns per timestamp attribute back! :-P

**Comment by mallinus** [comment]
On a serious note I don't mind making a PR?

**Comment by GrahamCampbell** [comment]
> It should return a DateTime instance

A carbon instance is a date time instance.

**Comment by mallinus** [comment]
Uh my bad it does indeed not matter what type is used when setting the columns for inserting in the database. However once the insert statement is done executing and once the model is returned, these Carbon attributes need to be casted to strings in case getDates() is an empty array, so that the $casts property can be used to transform (cast) the unix timestamps to integers for the API endpoint. When attributes are Carbon instances, an accessor is needed for each date attribute to get the timestamp in string in unix format.

Basically, this wont work:
`protected $casts = ['created_at' => 'integer', 'updated_at' => 'integer'];`
Unless:
`public function getCreatedAtAttribute($value)
    {
        $value = $value->timestamp;
    }`
